### PR TITLE
Project Summary Report

### PR DESCRIPTION
**Pull request by tmastrom**
Card: https://github.com/bcgov/cas-cif/issues/1061

### CODE REVIEW

File: schema/deploy/functions/handle_project_summary_report_form_change_commit.sql [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+            gross_amount = (fc.new_form_data->>'projectSummaryReportPayment')::numeric,
+            net_amount = (fc.new_form_data->>'projectSummaryReportPayment')::numeric,
+            date_sent_to_csnr = (fc.new_form_data->>'dateSentToCsnr')::timestamptz
+        where py.reporting_requirement_id = fc.form_data_record_id;
+
+    elsif fc.operation = 'archive' then
+
+        update cif.reporting_requirement set archived_at = now() where id = fc.form_data_record_id;
+        update cif.payment set archived_at = now() where reporting_requirement_id = fc.form_data_record_id;
+
```
Comment (Sepehr-Sobhani) [inline schema/deploy/functions/handle_project_summary_report_form_change_commit.sql:74]: We might want to have one for the project summary report as well if we decide to discard this report.

**Review by Sepehr-Sobhani — COMMENTED** [review]
Great stuff! 👌
I've added some comments but there's also one more. We need to add `...ProjectSummaryReportFormSummary_projectRevision` to `CollapsibleFormWidget.tsx` too in order to show Amendment/Revision changes properly.

File: app/components/Form/ProjectSummaryReportForm.tsx [hunk 2]
```diff
@@ -1,13 +1,196 @@
...
+        id
+        rowId
+        # eslint-disable-next-line relay/unused-fields
+        projectFormChange {
+          formDataRecordId
+        }
+        projectSummaryFormChanges: formChangesFor(
+          formDataTableName: "reporting_requirement"
+          reportType: "Project Summary Report"
+          first: 1000
```
Comment (pbastia) [inline app/components/Form/ProjectSummaryReportForm.tsx:35]: There should only be one project summary report, no need to fetch more I think

File: app/components/Form/ProjectSummaryReportForm.tsx [hunk 3]
```diff
@@ -1,13 +1,196 @@
...
+        upcomingProjectSummaryReportFormChange: upcomingReportingRequirementFormChange(
+          reportType: "Project Summary Report"
+        ) {
+          id
+          # eslint-disable-next-line relay/must-colocate-fragment-spreads
+          ...ReportDueIndicator_formChange
+          asReportingRequirement {
+            reportDueDate
+          }
+        }
```
Comment (pbastia) [inline app/components/Form/ProjectSummaryReportForm.tsx:58]: The 'upcoming' requirement will always return the same one as the previous one, so you can add the fields / fragment spread to the `projectSummaryFormChanges` just above instead

**Review by pbastia — CHANGES_REQUESTED** [review]
Nice work! Just started having a look, I'll finish tomorrow

File: app/components/Form/ProjectSummaryReportForm.tsx (hunk 3 above)
Comment (tmastrom) [reply app/components/Form/ProjectSummaryReportForm.tsx:58]: It looks like the this is the case after verifying with graphiql

**Review by pbastia — CHANGES_REQUESTED** [review]
A few more comments. Nice work!

File: app/components/Form/ProjectSummaryReportForm.tsx [hunk 4]
```diff
@@ -1,13 +1,196 @@
...
+              onChange={(change) => handleChange(change.formData)}
+              ref={(el) => (formRefs.current[formChangeNode.id] = el)}
+              formData={formChangeNode.newFormData}
+              ObjectFieldTemplate={EmptyObjectFieldTemplate}
+            />
+            <Button
+              size="medium"
+              variant="primary"
+              onClick={() =>
+                stageMultipleReportingRequirementFormChanges(
```
[the remark this answers is not in the document]
Comment (tmastrom) [reply app/components/Form/ProjectSummaryReportForm.tsx]: stageReportFormChanges in `reportingRequirementFormChangeFunctions.ts` has a comment saying it is deprecated. Should we write a new function for this?

[the remark this answers is not in the document]
Comment (pbastia) [reply app/components/Form/ProjectSummaryReportForm.tsx]: No, I think you can use the `stageFormChange` function directly here

File: app/components/Form/ProjectSummaryReportForm.tsx [hunk 5]
```diff
@@ -1,13 +1,199 @@
...
+            <UndoChangesButton
+              formChangeIds={[projectSummaryFormChange.rowId]}
+              formRefs={formRefs}
+            />
+            <SavingIndicator isSaved={!isUpdating && !isCreating} />
+          </header>
+          <Status
+            upcomingReportDueDate={upcomingReportDueDate}
+            reportSubmittedDates={reportSubmittedDates}
+            // TODO: fix status text (make singular??)
```
[the remark this answers is not in the document]
Comment (tmastrom) [reply app/components/Form/ProjectSummaryReportForm.tsx]: I'll move this to a card

File: app/cypress/support/commands.js [hunk 6]
```diff
@@ -258,6 +258,91 @@ Cypress.Commands.add(
...
+    // comments,
+    // projectSummaryReportPayment,
+    // paymentNotes,
+    // dateSentToCsnr
+  ) => {
+    cy.findByText(/Report Due Date$/i)
+      .next()
+      .contains(reportDueDate);
+
+    // cy.findByText(/Received Date$/i)
```
Comment (tmastrom) [inline app/cypress/support/commands.js:324]: It is a mystery to me why cypress cannot find these elements when they are present in the screenshots. Any input is welcome

File: app/components/ReportingRequirement/Status.tsx [hunk 7]
```diff
@@ -11,9 +11,14 @@ const Status: React.FC<Props> = ({
   upcomingReportDueDate,
   reportSubmittedDates,
 }) => {
+  let reportString = "Reports ";
+  if (reportType === "Project Summary") {
+    reportString = "Report ";
+  }
+
```
Comment (tmastrom) [inline app/components/ReportingRequirement/Status.tsx:18]: There will only ever be one Project Summary Report so I wanted the copy to reflect that.

File: app/cypress/support/commands.js (hunk 6 above)
Comment (Sepehr-Sobhani) [reply app/cypress/support/commands.js:324]: Could be a dot issue on date fields where the CI browser is different from your local browser. try something like:
`.contains(/Jun(\.)? 16, 1991/i);`

File: app/components/ReportingRequirement/Status.tsx (hunk 7 above)
Comment (pbastia) [reply app/components/ReportingRequirement/Status.tsx:18]: we try to use `const` where possible:

```suggestion
  const reportString = 
    reportType === "Project Summary" ? "Reports " : "Report ";
```

File: app/cypress/e2e/cif/project-revision/create-first-project-revision.cy.js [hunk 8]
```diff
@@ -466,6 +488,7 @@ describe("when creating a project, the project page", () => {
     cy.findByText(/Review and Submit information/i).click();
     cy.contains("Review and Submit Project").should("be.visible");
     cy.findByRole("button", { name: /^discard/i }).click();
+    cy.wait(1000);
```
Comment (pbastia) [inline app/cypress/e2e/cif/project-revision/create-first-project-revision.cy.js:491]: Can we wait on something better, like screen on text? We basically need to wait for the previous click to complete the re-render

File: app/cypress/support/commands.js [hunk 9]
```diff
@@ -258,6 +258,91 @@ Cypress.Commands.add(
...
+    //   .next()
+    //   .should("have.text", projectSummaryReportPayment);
+
+    // cy.findByText(/Notes for the Payment$/i)
+    //   .next()
+    //   .should("have.text", paymentNotes);
+
+    // cy.findByText(/Date Invoice Sent to CSNR$/i)
+    //   .next()
+    //   .contains(dateSentToCsnr);
```
Comment (pbastia) [inline app/cypress/support/commands.js:342]: This needs to be fixed before we can merge

**Review by pbastia — DISMISSED** [review]
Just a few more things. Nice work!

File: app/cypress/support/commands.js (hunk 6 above)
Comment (tmastrom) [reply app/cypress/support/commands.js]: I moved away from using the extending functions from `cypress testing library` to using the core `cypress` function. A more detailed description of these issues is [here](https://docs.google.com/document/d/14JlhKw9zxnD1GNN62XD-IOn2AL_ubPrW40I1F74wxxQ/edit#heading=h.9lzly6n5lxjr)

**Pull request merged**